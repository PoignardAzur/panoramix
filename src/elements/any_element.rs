use crate::ctx::{ProcessEventCtx, ReconcileCtx};
use crate::element_tree::{Element, VirtualDom};
use crate::glue::GlobalEventCx;
use crate::metadata::NoState;
use crate::widgets::WidgetSeqBox;

use std::any::Any;
use std::fmt::Debug;

// --- STATE ---

pub trait AnyState: Any + Debug {
    fn as_any(&self) -> &dyn Any;
    fn as_mut_any(&mut self) -> &mut dyn Any;

    fn print_type(&self) {
        println!("{:#?}", std::any::type_name::<Self>());
    }

    fn dyn_clone(&self) -> Box<dyn AnyState>;
    fn dyn_eq(&self, other: &Box<dyn AnyState>) -> bool;
}

impl<T> AnyState for T
where
    T: Clone + Default + Debug + PartialEq + 'static,
{
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_mut_any(&mut self) -> &mut dyn Any {
        self
    }

    fn dyn_clone(&self) -> Box<dyn AnyState> {
        Box::new(self.clone())
    }

    fn dyn_eq(&self, other: &Box<dyn AnyState>) -> bool {
        if let Some(other) = other.as_ref().as_any().downcast_ref::<Self>() {
            other.eq(self)
        } else {
            false
        }
    }
}

pub struct AnyStateBox {
    value: Box<dyn AnyState>,
}

impl AnyStateBox {
    pub fn new(state: impl AnyState) -> Self {
        AnyStateBox {
            value: Box::new(state),
        }
    }
}

impl Clone for AnyStateBox {
    fn clone(&self) -> Self {
        AnyStateBox {
            value: self.value.dyn_clone(),
        }
    }
}

impl Debug for AnyStateBox {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.value.fmt(f)
    }
}

impl PartialEq for AnyStateBox {
    fn eq(&self, other: &Self) -> bool {
        self.value.dyn_eq(&other.value)
    }
}

// --- ELEMENT ---

#[derive(Default, Clone, PartialEq, Eq, Hash)]
struct ErasedElement<Child: Element> {
    child: Option<Child>,
}

impl<Child: Element> Debug for ErasedElement<Child> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.child.as_ref().unwrap().fmt(f)
    }
}

trait AnyElement: Any + Debug {
    type Event;

    fn print_type(&self) {
        println!("{:#?}", std::any::type_name::<Self>());
    }

    fn dyn_clone(&self) -> Box<dyn AnyElement<Event = Self::Event>>;

    fn build(
        &mut self,
        prev_state: Option<AnyStateBox>,
    ) -> (
        Box<dyn AnyVirtualDom<Event = Self::Event>>,
        Option<AnyStateBox>,
    );
}

impl<Child: Element + 'static> AnyElement for ErasedElement<Child> {
    type Event = Child::Event;

    fn dyn_clone(&self) -> Box<dyn AnyElement<Event = Self::Event>> {
        Box::new(self.clone())
    }

    fn build(
        &mut self,
        prev_state: Option<AnyStateBox>,
    ) -> (
        Box<dyn AnyVirtualDom<Event = Self::Event>>,
        Option<AnyStateBox>,
    ) {
        let child = self.child.take().unwrap();

        let prev_state = if let Some(mut prev_state) = prev_state {
            std::mem::take(
                prev_state
                    .value
                    .as_mut_any()
                    .downcast_mut::<Child::AggregateChildrenState>()
                    .unwrap(),
            )
        } else {
            Default::default()
        };

        let (output, state) = child.build(prev_state);

        (
            Box::new(ErasedVirtualDom { child: output }),
            Some(AnyStateBox {
                value: Box::new(state),
            }),
        )
    }
}

// -

pub struct ElementBox<Event> {
    child: Box<dyn AnyElement<Event = Event>>,
}

impl<Event> ElementBox<Event> {
    pub fn new(child: impl Element<Event = Event> + 'static) -> Self {
        ElementBox {
            child: Box::new(ErasedElement { child: Some(child) }),
        }
    }
}

impl<Event> Debug for ElementBox<Event> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.child.fmt(f)
    }
}

impl<Event> Clone for ElementBox<Event> {
    fn clone(&self) -> Self {
        ElementBox {
            child: self.child.dyn_clone(),
        }
    }
}

impl<Event> Element for ElementBox<Event> {
    type Event = Event;
    type ComponentState = NoState;
    type AggregateChildrenState = Option<AnyStateBox>;
    type BuildOutput = VirtualDomBox<Event>;

    fn build(self, prev_state: Option<AnyStateBox>) -> (VirtualDomBox<Event>, Option<AnyStateBox>) {
        let mut child = self.child;
        let (output, state) = child.build(prev_state);

        (VirtualDomBox { child: output }, state)
    }
}

// --- VIRTUAL_DOM ---

#[derive(Default, Clone, PartialEq, Eq, Hash)]
pub struct ErasedVirtualDom<Child: VirtualDom> {
    child: Child,
}

impl<Child: VirtualDom> Debug for ErasedVirtualDom<Child> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.child.fmt(f)
    }
}

pub trait AnyVirtualDom: Any + Debug {
    type Event;

    fn as_any(&self) -> &dyn Any;

    fn print_type(&self) {
        println!("{:#?}", std::any::type_name::<Self>());
    }

    fn init_tree(&self) -> WidgetSeqBox;

    fn reconcile(
        &self,
        other: &Box<dyn AnyVirtualDom<Event = Self::Event>>,
        widget_seq: &mut WidgetSeqBox,
        ctx: &mut ReconcileCtx,
    );

    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    );

    fn process_local_event(
        &self,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event>;
}

impl<Child: VirtualDom + 'static> AnyVirtualDom for ErasedVirtualDom<Child> {
    type Event = Child::Event;

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn init_tree(&self) -> WidgetSeqBox {
        WidgetSeqBox {
            value: Box::new(self.child.init_tree()),
        }
    }

    fn reconcile(
        &self,
        other: &Box<dyn AnyVirtualDom<Event = Self::Event>>,
        widget_seq: &mut WidgetSeqBox,
        ctx: &mut ReconcileCtx,
    ) {
        let other = other.as_any().downcast_ref::<Self>().unwrap();
        let widget_seq = widget_seq
            .value
            .as_mut_any()
            .downcast_mut::<Child::TargetWidgetSeq>()
            .unwrap();
        self.child.reconcile(&other.child, widget_seq, ctx);
    }

    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    ) {
        let children_state = children_state
            .as_mut()
            .unwrap()
            .value
            .as_mut_any()
            .downcast_mut::<Child::AggregateChildrenState>()
            .unwrap();
        let widget_seq = widget_seq
            .value
            .as_mut_any()
            .downcast_mut::<Child::TargetWidgetSeq>()
            .unwrap();
        self.child
            .process_event(comp_ctx, children_state, widget_seq, cx)
    }

    fn process_local_event(
        &self,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event> {
        let children_state = children_state
            .as_mut()
            .unwrap()
            .value
            .as_mut_any()
            .downcast_mut::<Child::AggregateChildrenState>()
            .unwrap();
        let widget_seq = widget_seq
            .value
            .as_mut_any()
            .downcast_mut::<Child::TargetWidgetSeq>()
            .unwrap();
        self.child
            .process_local_event(children_state, widget_seq, cx)
    }
}

// -

pub struct VirtualDomBox<Event> {
    child: Box<dyn AnyVirtualDom<Event = Event>>,
}

impl<Event> VirtualDomBox<Event> {
    pub fn new(child: impl VirtualDom<Event = Event> + 'static) -> Self {
        VirtualDomBox {
            child: Box::new(ErasedVirtualDom { child: child }),
        }
    }
}

impl<Event> Debug for VirtualDomBox<Event> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.child.fmt(f)
    }
}

impl<Event> VirtualDom for VirtualDomBox<Event> {
    // TODO - Accept any Event parent of Self::Event
    type Event = Event;
    type AggregateChildrenState = Option<AnyStateBox>;
    type TargetWidgetSeq = WidgetSeqBox;

    fn init_tree(&self) -> Self::TargetWidgetSeq {
        self.child.init_tree()
    }

    fn reconcile(
        &self,
        other: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        self.child.reconcile(&other.child, widget_seq, ctx);
    }

    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    ) {
        self.child
            .process_event(comp_ctx, children_state, widget_seq, cx)
    }

    fn process_local_event(
        &self,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event> {
        self.child
            .process_local_event(children_state, widget_seq, cx)
    }
}

// --- TESTS ---

#[cfg(test)]
mod tests {
    use super::*;
    use crate::elements::{Button, Label};
    use crate::test_harness::Harness;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn any_state() {
        let state1 = AnyStateBox::new(42);
        let state2 = state1.clone();

        assert!(state1 == state2);
    }

    #[test]
    fn new_element() {
        let label = ElementBox::new(Label::new("Hello"));
        assert_debug_snapshot!(label);

        let (label_data, _state) = label.build(None);
        assert_debug_snapshot!(label_data);

        // TODO - check state
    }

    #[test]
    fn boxed_label_widget() {
        let label = ElementBox::new(Label::new("Hello"));

        Harness::run_test_window(label, |harness| {
            let label_state = harness.get_root_debug_state();
            assert_debug_snapshot!(label_state);

            let new_label = ElementBox::new(Label::new("World"));
            harness.update_root_element(new_label);

            let label_state_2 = harness.get_root_debug_state();
            assert_debug_snapshot!(label_state_2);
        });
    }

    // FIXME - Test doesn't use box
    #[test]
    fn boxed_button_press() {
        use crate::elements::event_logger::EventLogger;
        use crate::glue::WidgetId;
        use std::sync::mpsc::channel;

        let (event_sender, event_receiver) = channel();
        let button_id = WidgetId::reserved(1);
        let button = EventLogger::new(
            event_sender,
            Button::new("Hello").with_reserved_id(button_id),
        );

        Harness::run_test_window(button, |harness| {
            assert_debug_snapshot!(harness.get_root_debug_state());

            harness.mouse_click_on(button_id);

            let click_event = event_receiver.try_recv();
            assert_debug_snapshot!(click_event);
        });
    }
}
