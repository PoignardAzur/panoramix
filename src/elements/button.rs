use crate::glue::{Action, GlobalEventCx, WidgetId};

use crate::element_tree::{Element, ElementExt, Metadata, NoState, VirtualDom};
use crate::flex::FlexParams;
use crate::widgets::ButtonWidget;

use crate::element_tree::ReconcileCtx;

use derivative::Derivative;
use tracing::{instrument, trace};

/// A button with a text label.
///
/// ## Events
///
/// Emits [ButtonClick] events.
#[derive(Derivative, PartialEq)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = ""))]
pub struct Button {
    pub text: String,
    pub flex: FlexParams,
    pub reserved_widget_id: Option<WidgetId>,
}

#[derive(Derivative, PartialEq)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = ""))]
pub struct ButtonData {
    pub text: String,
    pub flex: FlexParams,
    pub reserved_widget_id: Option<WidgetId>,
}

/// Event emitted when a [Button] is clicked.
///
/// Note: Might hold data like "mouse position" or "button id" future versions.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct ButtonClick;

//
// --- IMPLS

impl Button {
    /// Build a button with the given label.
    ///
    /// Use the [.on_click](Button::on_click) method to provide a closure to be called when the button is clicked.
    pub fn new(text: impl Into<String>) -> Self {
        Button {
            text: text.into(),
            flex: FlexParams {
                flex: 1.0,
                alignment: None,
            },
            reserved_widget_id: None,
        }
    }

    /// Change the way the button's size is calculated
    pub fn with_flex_params(self, flex_params: FlexParams) -> Self {
        Button {
            flex: flex_params,
            ..self
        }
    }

    // TODO - doc
    // For unit tests only
    pub fn with_reserved_id(self, widget_id: WidgetId) -> Self {
        Button {
            reserved_widget_id: Some(widget_id),
            ..self
        }
    }

    /// Provide a closure to be called when this button is clicked.
    pub fn on_click<ComponentEvent: 'static, ComponentState: 'static>(
        self,
        md: Metadata<ComponentEvent, ComponentState>,
        callback: impl Fn(&mut ComponentState, ButtonClick) + Clone,
    ) -> impl Element {
        self.on(md, callback)
    }
}

impl Element for Button {
    type Event = ButtonClick;

    type ComponentState = NoState;
    type AggregateChildrenState = ();
    type BuildOutput = ButtonData;

    #[instrument(name = "Button", skip(self, _prev_state))]
    fn build(self, _prev_state: ()) -> (ButtonData, ()) {
        (
            ButtonData {
                text: self.text,
                flex: self.flex,
                reserved_widget_id: self.reserved_widget_id,
            },
            (),
        )
    }
}

impl VirtualDom for ButtonData {
    type Event = ButtonClick;
    type AggregateChildrenState = ();
    type TargetWidgetSeq = ButtonWidget;

    #[instrument(name = "Button", skip(self))]
    fn init_tree(&self) -> ButtonWidget {
        let id = self.reserved_widget_id.unwrap_or_else(WidgetId::next);
        ButtonWidget::new(self.text.clone(), self.flex, id)
    }

    #[instrument(name = "Button", skip(self, _other, _widget, _ctx))]
    fn reconcile(&self, _other: &Self, _widget: &mut ButtonWidget, _ctx: &mut ReconcileCtx) {
        //widget.set_text(self.text.clone());
    }

    #[instrument(name = "Button", skip(self, _children_state, widget, cx))]
    fn process_local_event(
        &self,
        _children_state: &mut Self::AggregateChildrenState,
        widget: &mut ButtonWidget,
        cx: &mut GlobalEventCx,
    ) -> Option<ButtonClick> {
        // FIXME - Rework event dispatching
        let id = widget.id();
        if let Some(Action::Clicked) = cx.app_data.dequeue_action(id) {
            trace!("Processed button press");
            Some(ButtonClick)
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::element_tree::assign_empty_state_type;
    use crate::test_harness::Harness;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn new_button() {
        let button = Button::new("Hello");
        let (button_data, ()) = button.clone().build(());

        assert_debug_snapshot!(button);
        assert_debug_snapshot!(button_data);

        assert_eq!(
            button_data,
            ButtonData {
                text: String::from("Hello"),
                flex: FlexParams {
                    flex: 1.0,
                    alignment: None,
                },
                ..Default::default()
            }
        );

        assign_empty_state_type(&button);
    }

    #[test]
    fn button_widget() {
        // TODO - We use Tuple! because RootWidget currently wants a root element with no event
        use crate::Tuple;
        let button = Tuple!(Button::new("Hello"));

        Harness::run_test_window(button, |harness| {
            let button_state = harness.get_root_debug_state();
            assert_debug_snapshot!(button_state);

            // TODO - Test reconcile() method (currently doesn't work)
        });
    }

    #[test]
    fn button_press() {
        use crate::elements::event_logger::EventLogger;
        use std::sync::mpsc::channel;

        let (event_sender, event_receiver) = channel();
        let button_id = WidgetId::reserved(1);
        let button = EventLogger::new(
            event_sender,
            Button::new("Hello").with_reserved_id(button_id),
        );

        Harness::run_test_window(button, |harness| {
            assert_debug_snapshot!(harness.get_root_debug_state());

            harness.mouse_click_on(button_id);

            let click_event = event_receiver.try_recv();
            assert_debug_snapshot!(click_event);
        });
    }
}
