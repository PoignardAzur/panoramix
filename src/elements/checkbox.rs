use crate::glue::{Action, GlobalEventCx, WidgetId};

use crate::element_tree::{Element, ElementExt, NoEvent, VirtualDom};
use crate::flex::FlexParams;
use crate::widgets::{CheckboxWidget, SingleCheckboxWidget};

use crate::element_tree::ReconcileCtx;

use derivative::Derivative;
use tracing::{instrument, trace};

// TODO - Handle the anti-pattern where the user does something like
// Checkbox::new(false)
// In other words, enforce two-ways bindings

/// A checkbox with a text label.
///
/// ## Events
///
/// Emits [Toggled] events.
#[derive(Derivative, PartialEq)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = ""))]
pub struct Checkbox<CpEvent = NoEvent, CpState = ()> {
    pub text: String,
    pub value: bool,
    pub flex: FlexParams,
    pub reserved_widget_id: Option<WidgetId>,
    #[derivative(Debug = "ignore")]
    pub _markers: std::marker::PhantomData<(CpEvent, CpState)>,
}

#[derive(Derivative, PartialEq)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = ""))]
pub struct CheckboxData<CpEvent = NoEvent, CpState = ()> {
    pub text: String,
    pub value: bool,
    pub flex: FlexParams,
    pub reserved_widget_id: Option<WidgetId>,
    #[derivative(Debug = "ignore")]
    pub _markers: std::marker::PhantomData<(CpEvent, CpState)>,
}

/// Event emitted when a [Checkbox] is clicked.
///
/// Note: Might hold data like "mouse position" or "checkbox id" future versions.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct Toggled {
    pub new_value: bool,
}

//
// --- IMPLS

impl<CpEvent, CpState> Checkbox<CpEvent, CpState> {
    /// Build a checkbox with the given label.
    ///
    /// Use the [.on_toggled](Checkbox::on_toggled) method to provide a closure to be called when the box is toggled.
    pub fn new(text: impl Into<String>, value: bool) -> Self {
        Checkbox {
            text: text.into(),
            value,
            flex: FlexParams {
                flex: 1.0,
                alignment: None,
            },
            reserved_widget_id: None,
            _markers: Default::default(),
        }
    }

    /// Change the way the checkbox's size is calculated
    pub fn with_flex_params(self, flex_params: FlexParams) -> Self {
        Checkbox {
            flex: flex_params,
            ..self
        }
    }

    pub fn with_reserved_id(self, widget_id: WidgetId) -> Self {
        Checkbox {
            reserved_widget_id: Some(widget_id),
            ..self
        }
    }

    /// Provide a closure to be called when this checkbox is toggled.
    pub fn on_toggled(
        self,
        callback: impl Fn(&mut CpState, Toggled) + Clone,
    ) -> impl Element<CpEvent, CpState> {
        self.on(callback)
    }
}

impl<CpEvent, CpState> Element<CpEvent, CpState> for Checkbox<CpEvent, CpState> {
    type Event = Toggled;
    type AggregateChildrenState = ();
    type BuildOutput = CheckboxData<CpEvent, CpState>;

    #[instrument(name = "Checkbox", skip(self, _prev_state))]
    fn build(self, _prev_state: ()) -> (CheckboxData<CpEvent, CpState>, ()) {
        (
            CheckboxData {
                text: self.text,
                value: self.value,
                flex: self.flex,
                reserved_widget_id: self.reserved_widget_id,
                _markers: Default::default(),
            },
            (),
        )
    }
}

impl<CpEvent, CpState> VirtualDom<CpEvent, CpState> for CheckboxData<CpEvent, CpState> {
    type Event = Toggled;
    type AggregateChildrenState = ();

    type TargetWidgetSeq = SingleCheckboxWidget;

    #[instrument(name = "Checkbox", skip(self))]
    fn init_tree(&self) -> SingleCheckboxWidget {
        let id = self.reserved_widget_id.unwrap_or_else(WidgetId::next);
        SingleCheckboxWidget::new(
            CheckboxWidget::new(self.text.clone(), self.value, id),
            self.flex,
        )
    }

    #[instrument(name = "Checkbox", skip(self, other, widget, ctx))]
    fn reconcile(&self, other: &Self, widget: &mut SingleCheckboxWidget, ctx: &mut ReconcileCtx) {
        let checkbox_widget = widget.widget_mut();
        if self.text != other.text {
            // TODO
            //checkbox_widget.pod.widget_mut().set_text(self.text.clone());
        }
        checkbox_widget.value = self.value;
        // TODO - check diff with previous value
        widget.request_druid_update(ctx);
        widget.widget_mut().request_druid_update(ctx);
    }

    #[instrument(name = "Checkbox", skip(self, _children_state, widget, cx))]
    fn process_local_event(
        &self,
        _children_state: &mut Self::AggregateChildrenState,
        widget: &mut SingleCheckboxWidget,
        cx: &mut GlobalEventCx,
    ) -> Option<Toggled> {
        // FIXME - Rework event dispatching
        let id = widget.widget().id();
        if let Some(Action::Clicked) = cx.app_data.dequeue_action(id) {
            let new_value = widget.widget().value;
            trace!("Processed checkbox toggle: {}", new_value);
            Some(Toggled { new_value })
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::element_tree::assign_empty_state_type;
    use crate::test_harness::Harness;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    // TODO - Test flex params?

    #[test]
    fn new_checkbox() {
        let checkbox = Checkbox::new("Hello", false);
        let (checkbox_data, ()) = checkbox.clone().build(());

        assert_debug_snapshot!(checkbox);
        assert_debug_snapshot!(checkbox_data);

        assert_eq!(
            checkbox_data,
            CheckboxData {
                text: String::from("Hello"),
                value: false,
                flex: FlexParams {
                    flex: 1.0,
                    alignment: None,
                },
                ..Default::default()
            }
        );

        assign_empty_state_type(&checkbox);
    }

    #[test]
    fn checkbox_widget() {
        // TODO - We use Tuple! because RootWidget currently wants a root element with no event
        use crate::Tuple;
        let checkbox = Tuple!(Checkbox::new("Hello", false));

        Harness::run_test_window(checkbox, |harness| {
            let checkbox_state = harness.get_root_debug_state();
            assert_debug_snapshot!(checkbox_state);

            let new_checkbox = Tuple!(Checkbox::new("Hello", true));
            harness.update_root_element(new_checkbox);

            let checkbox_state_2 = harness.get_root_debug_state();
            assert_debug_snapshot!(checkbox_state_2);

            // TODO - Test reconcile() with text (currently not implemented for checkbox)
        });
    }

    // FIXME - Test currently doesn't work; checkbox doesn't emit events
    #[allow(dead_code)]
    fn checkbox_press() {
        use crate::elements::event_logger::EventLogger;
        use std::sync::mpsc::channel;

        let (event_sender, event_receiver) = channel();
        let checkbox_id = WidgetId::reserved(1);
        let checkbox = EventLogger::new(
            event_sender,
            Checkbox::new("Hello", false).with_reserved_id(checkbox_id),
        );

        Harness::run_test_window(checkbox, |harness| {
            assert_debug_snapshot!(harness.get_root_debug_state());

            harness.mouse_click_on(checkbox_id);

            let click_event = event_receiver.try_recv();
            assert_debug_snapshot!(click_event);

            // TODO - test data persistence, somehow?
        });
    }
}
