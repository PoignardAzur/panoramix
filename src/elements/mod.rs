//! GUI elements that can be built in a component.

mod any_element;
mod button;
mod checkbox;
mod component;
mod element_list;
mod element_option;
mod empty;
mod flex_element;
mod label;
mod textbox;

pub mod element_tuple;
mod event_logger;
mod mock_component;
mod with_event;

mod compute_diff;

pub use any_element::ElementBox;
pub use button::{Button, ButtonClick};
pub use checkbox::{Checkbox, Toggled};
pub use component::{Component, ComponentOutput};
pub use element_list::ElementList;
pub use element_option::*;
pub use empty::EmptyElement;
pub use flex_element::Flex;
pub use label::Label;
pub use textbox::{TextBox, TextChanged};

// TODO - doc
pub mod internals {
    pub use super::any_element::VirtualDomBox;
    pub use super::button::ButtonData;
    pub use super::checkbox::CheckboxData;
    pub use super::component::{ComponentHolder, ComponentOutputData};
    pub use super::element_list::ElementListData;
    pub use super::element_tuple::ElementTupleData;
    pub use super::empty::EmptyElementData;
    pub use super::flex_element::FlexData;
    pub use super::label::LabelData;
    pub use super::textbox::TextBoxData;
    pub use super::with_event::WithEventTarget;
    pub use super::with_event::{ParentEvent, WithBubbleEvent, WithCallbackEvent, WithMapEvent};

    // TODO - move to test_harness?
    pub use super::event_logger::{EventLogger, EventLoggerData};
    pub use super::mock_component::{MockComponent, MockComponentData, MockState};

    pub use super::compute_diff::{compute_diff, ListMutation, ListMutationItem};
}
