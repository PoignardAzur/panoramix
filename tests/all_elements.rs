use panoramix::elements::{
    Button, Checkbox, ComponentOutput, ElementBox, ElementList, Label, TextBox, TextChanged,
    Toggled,
};
use panoramix::internals::WidgetId;
use panoramix::{component, Column, CompCtx, Element, ElementExt, NoEvent, Row};

use either::{Left, Right};

#[derive(Debug, Default, Clone, PartialEq)]
struct TaskItem {
    text: String,
    is_completed: bool,
    id: u16,
}

#[derive(Debug, Clone, PartialEq)]
struct AppState {
    tasks: Vec<TaskItem>,
    high_priority: bool,
    task_name: String,
    next_id: u16,
}

// We implement the initial state of the app
impl Default for AppState {
    fn default() -> Self {
        let tasks: Vec<_> = (0..4)
            .into_iter()
            .map(|i| TaskItem {
                text: format!("Task #{}", i),
                is_completed: i >= 2,
                id: i,
            })
            .collect();
        let next_id = tasks.len() as u16;

        AppState {
            tasks,
            high_priority: false,
            task_name: String::from("Next task"),
            next_id,
        }
    }
}

type ItemEvent = Toggled;

#[component]
fn TaskRow(ctx: &CompCtx, props: TaskItem) -> impl Element<Event = ItemEvent> {
    let md = ctx.use_metadata::<ItemEvent, ()>();
    let text = props.text.clone();

    let checkbox = ElementBox::new(
        Checkbox::new("", props.is_completed).with_reserved_id(WidgetId::reserved(props.id + 10)),
    )
    .bubble_up::<ItemEvent, _, _>(md);
    let row = if props.is_completed {
        Left(Row!(checkbox, Label::new(text)))
    } else {
        Right(Row!(checkbox, Label::new(text), Label::new("(complete)")))
    };

    ComponentOutput::new(md, row)
}

#[component]
fn AwesomeEditableList(ctx: &CompCtx, _props: ()) -> impl Element<Event = NoEvent> {
    let md = ctx.use_metadata::<NoEvent, AppState>();
    let state = ctx.get_local_state(md);

    // TODO - Add "validate on enter" feature
    let textbox_task_name = TextBox::new(state.task_name.clone()).on_text_changed(
        md,
        |state: &mut AppState, event: TextChanged| {
            state.task_name = event.new_content;
        },
    );

    let button_new_task = Button::new("New task")
        .with_reserved_id(WidgetId::reserved(1))
        .on_click(md, |state: &mut AppState, _| {
            if state.task_name == "" {
                return;
            }

            let new_task = TaskItem {
                text: state.task_name.clone(),
                is_completed: false,
                id: state.next_id,
            };

            state.tasks.push(new_task);
            state.task_name = String::new();
            state.next_id += 1;
        });

    // TODO - Find a syntax that looks more hierachical
    let list_keys = state.tasks.iter().map(|task_item| task_item.id.to_string());
    let list_rows = state.tasks.iter().enumerate().map(|(i, task_item)| {
        TaskRow::new(task_item.clone()).on::<ItemEvent, _, _, _>(
            md,
            move |state: &mut AppState, event| {
                state.tasks[i].is_completed = event.new_value;
            },
        )
    });
    let list_view = ElementList::from_keys_elems(list_keys, list_rows);

    let button_delete = Button::new("Delete completed tasks")
        .with_reserved_id(WidgetId::reserved(2))
        .on_click(md, |state: &mut AppState, _| {
            state.tasks = std::mem::take(&mut state.tasks)
                .into_iter()
                .filter(|task| !task.is_completed)
                .collect();
        });

    ComponentOutput::new(
        md,
        Column!(
            Row!(textbox_task_name, button_new_task),
            list_view,
            button_delete,
        ),
    )
}

use insta::assert_debug_snapshot;
use panoramix::test_harness::Harness;
use test_env_log::test;

#[test]
fn test_all_widgets() {
    let list = AwesomeEditableList::new(());

    let button_new_task_id = WidgetId::reserved(1);
    let button_delete_id = WidgetId::reserved(2);
    //let checkbox_0_id = WidgetId::reserved(10);
    //let checkbox_2_id = WidgetId::reserved(12);

    Harness::run_test_window(list, |harness| {
        let root_state = harness.get_root_debug_state();
        assert_debug_snapshot!(root_state);

        harness.mouse_click_on(button_new_task_id);

        let root_state_2 = harness.get_root_debug_state();
        assert_debug_snapshot!(root_state_2);

        // FIXME - Add "edit textbox" event
        // FIXME - Add "click checkbox" event

        //harness.mouse_click_on(checkbox_0_id);
        //harness.mouse_click_on(checkbox_2_id);

        //let root_state_3 = harness.get_root_debug_state();
        //assert_debug_snapshot!(root_state_3);

        harness.mouse_click_on(button_delete_id);

        let root_state_4 = harness.get_root_debug_state();
        assert_debug_snapshot!(root_state_4);
    });
}
