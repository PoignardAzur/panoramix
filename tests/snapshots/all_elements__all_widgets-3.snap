---
source: tests/all_elements.rs
expression: root_state_4

---
RootWidget(
    Column(
        Column(
            Row(
                IdentityWrapper(
                    TextBox,
                ),
                IdentityWrapper(
                    ControllerHost(
                        Button(
                            "New task",
                        ),
                    ),
                ),
            ),
            Row(
                IdentityWrapper(
                    Checkbox(
                        "[_] ",
                    ),
                ),
                Label(
                    "Task #0",
                ),
                Label(
                    "(complete)",
                ),
            ),
            Row(
                IdentityWrapper(
                    Checkbox(
                        "[_] ",
                    ),
                ),
                Label(
                    "Task #1",
                ),
                Label(
                    "(complete)",
                ),
            ),
            Row(
                IdentityWrapper(
                    Checkbox(
                        "[_] ",
                    ),
                ),
                Label(
                    "Next task",
                ),
                Label(
                    "(complete)",
                ),
            ),
            IdentityWrapper(
                ControllerHost(
                    Button(
                        "Delete completed tasks",
                    ),
                ),
            ),
        ),
    ),
)
